feat(session): allow multiple root domains

mind-ar opened the pull request:
Hi, this PR implements  #1198.
Personally I don't like the result too much as I had to touch a lot of files, but it can give an idea to the maintainers.

```yml
session:
#  domain: example.com
  domain_list:
    - example.com
    - example2.com
  expiration: 3600  # 1 hour
  inactivity: 300  # 5 minutes
  remember_me_duration: 1y
```
Tasks:

- [x] create new suite: `MultipleDomain`
- [x] add new parameter in configuration
- [x] allow multiple root domains
- [ ] review / discuss / discard
- [ ] analyze possible security implications 
- [ ] improve code
- [ ] update docs
- [ ] write tests

Feel free to request any change

mind-ar commented:
I think that #3201 can be implemented too as part of this PR

james-d-elliott commented:
> I think that #3201 can be implemented too as part of this PR

Yep, The way to do this would be to implement the non-shared portion of sessions (i.e. everything other than redis and the secret) as it's own struct, 

I had in mind something like this:

```go
// SessionConfiguration represents the configuration related to user sessions.
type SessionConfiguration struct {
	Name               string        `koanf:"name"`
	Domain             string        `koanf:"domain"`
	SameSite           string        `koanf:"same_site"`
	Secret             string        `koanf:"secret"`
	Expiration         time.Duration `koanf:"expiration"`
	Inactivity         time.Duration `koanf:"inactivity"`
	RememberMeDuration time.Duration `koanf:"remember_me_duration"`

	Redis *RedisSessionConfiguration `koanf:"redis"`
	
	Domains []SessionDomainConfiguration  `koanf:"domains"`
}

type SessionDomainConfiguration struct {
	Name               string        `koanf:"name"`
	PortalDomain    string        `koanf:"domain"`
	CookieDomain  string        `koanf:"domain"`
	SameSite           string        `koanf:"same_site"`
	Expiration         time.Duration `koanf:"expiration"`
	Inactivity         time.Duration `koanf:"inactivity"`
	RememberMeDuration time.Duration `koanf:"remember_me_duration"`
}
```

mind-ar commented:
Hi @james-d-elliott 
so, the configuration would be something like that?
```yml
session:
  domain: example.com #backward compatibility
  domains:
    - example.com:
      expiration: 1800
      remember_me_duration: 3y 
    - example2.com:
      expiration: 600
      inactivity: 60
      remember_me_duration: 2d
  # default values
  same_site: true
  expiration: 3600  # 1 hour
  inactivity: 300  # 5 minutes
  remember_me_duration: 1y
```
I don't understand the difference between the `ALL DOMAINS` in item 1 and `Name` and `Domain` in point 3, are they the same?
I don't know if allowing the use of an unlisted domain can leaad to a security issue.

Note that the way I implemented the functionality, each domain or subdomain configured in the list, will have different session. for example, if i define `example.com` and `site1.example.com` they will treated as different domains (requesting login for both), so i don't think it can solve #3201 without doing dome ' magic' 

Before continuing I will wait for your opinion about how I implemented the functionality, maybe you have another idea of how to implement it. 
I'm going to summarize the changes below:
* modified nginx configuration to allow multiple domains in suites
* added a domain list in configuration
* in session provider, refactored  `*fasthttpsession.Session` as a `map[string]*fasthttpsession.Session`. the domain is the index
* for each request, i search the requested domain y `config.DomainList` and `config.Domain` (for backward compatibility) and get the user session for that domain

james-d-elliott commented:
> I don't understand the difference between the `ALL DOMAINS` in item 1 and `Name` and `Domain` in point 3, are they the same?
> I don't know if allowing the use of an unlisted domain can lead to a security issue.

There are some very low security attack vectors  from allowing an unknown domain. They're basically not an issue for us, and a lot would have to be configured incorrectly for it to have even the potential for it to be an issue. Ensuring we only respond on known domains can mitigate this even further than we do now.

It's also going to drastically simplify matching the correct cookie domain to the correct URL as well as the safe redirection check. The simpler they are the better for security.

> so, the configuration would be something like that?

More similar to this (maybe we make portal_domain optional and if not specified it is the cookie_domain):

```yaml
session:
  domain: example.com #backward compatibility
  domains:
    -   portal_domain: auth.example.com
        cookie_domain: example.com
        expiration: 1800
        remember_me_duration: 3y 
    -   portal_domain: auth.example2.com
        cookie_domain: example2.com
        expiration: 1800
        remember_me_duration: 3y 
  # default values
  same_site: true
  expiration: 3600  # 1 hour
  inactivity: 300  # 5 minutes
  remember_me_duration: 1y
```

> Before continuing I will wait for your opinion about how I implemented the functionality, maybe you have another idea of how to implement it.

I have always thought re-implementation of the session library for our specific needs was going to be the ideal solution. The current library was purpose built with a single domain in mind, not multiple. I'll think on it and see if I can think of anything specific when not doing re-implementation. 

> so i don't think it can solve #3201 without doing some ' magic'

Yeah you may be right. The question is how do we do this as well as #3201 without making configuration confusing and without making the code overly complicated?

mind-ar commented:
Hi @james-d-elliott , the `portal_domain` is the index we will use to get session? if yes, in the example bellow `auth.example.com` and `secure.example.com` will be on different session (user must login twice) but using the same cookie and things may get weird (unless we use different cookie for each domain?)
```yml
session:
  domains:
    -   portal_domain: auth.example.com
        cookie_domain: example.com
        expiration: 1800
        remember_me_duration: 3y 
    -   portal_domain: secure.example.com
        cookie_domain: example.com
        expiration: 1800
        remember_me_duration: 3y 
```
i think (correct me if i'm wrong) the portal domain and cookie domain must be always the same.

nightah commented:
I also think `portal_domain` is slightly confusing in terms of its name and what it's there for—especially considering use cases where the portal domain may not be a separate sub-domain to the protected resource (path-based Authelia).

I think we have something like `domain` and `cookie_domain` where the former is a list of domains that should be associated with the respective `cookie_domain` and other respective settings.

What do you guys think?

mind-ar commented:
@nightah  something like this?
```yml
sessions:
  - name: cookie_name
    domains:
      - auth.internal.example.com
      - internal.example.com
    cookie_domain: internal.example.com 
    expiration: 600  # 10 minutes
    inactivity: 180  # 3 minutes
    remember_me_duration: 1y 
  - name: cookie_name
    domains:
      - auth.example.com
      - secure.example.com
      - public.example.com
    cookie_domain: example.com 
    expiration: 3600  # 1 hour
    inactivity: 300  # 5 minutes
    remember_me_duration: 1y 
 - name: cookie_name
   domains:
      - example2.com
   cookie_domain: example2.com 
   expiration: 3600  # 1 hour
   inactivity: 300  # 5 minutes
   remember_me_duration: 1y 
```
@james-d-elliott what is `portal_domain`? is the authelia (sub)domain (ie `auth.example.com`)? why we need it? sorry if i'm lost

nightah commented:
Yep mostly like that. I believe we can drop the `name` option in that example though.

mind-ar commented:
Hi, will try to summarize the two alternatives

Alternative 1
-----
```yml
session:
  domain: example.com #backward compatibility
  domains:
    -   domain: internal.example.com # cookie domain
        expiration: 1800
        inactivity: 300  # 5 minutes
        remember_me_duration: 3y 
    -   domain: example.com # cookie domain
        expiration: 1800
        inactivity: 300  # 5 minutes
        remember_me_duration: 3y 
    -   domain: example2.com # cookie domain
        expiration: 1800
        inactivity: 300  # 5 minutes
        remember_me_duration: 3y 
  # default values
  same_site: true
  expiration: 3600  # 1 hour
  inactivity: 300  # 5 minutes
  remember_me_duration: 1y
```
**Notes**
* is backward compatible (`domain` can be used with an "deprecation warning") 
* less complicated configuration (?)
* match domain and subdomains (ie `internal.example.com` matchs `*.internal.example.com`)
* order matters (first match, like acl rules)
* provides default values

Alternative 2
----

```yml
sessions:
  - cookie_domain: internal.example.com
    domains:
      - auth.internal.example.com
      - internal.example.com
    cookie_domain: internal.example.com 
    expiration: 600  # 10 minutes
    inactivity: 180  # 3 minutes
    remember_me_duration: 1y 
  - cookie_domain: example.com
    domains:
      - auth.example.com
      - secure.example.com
      - public.example.com
    expiration: 3600  # 1 hour
    inactivity: 300  # 5 minutes
    remember_me_duration: 1y 
 - cookie_domain: example2.com
   domains:
      - example2.com
   expiration: 3600  # 1 hour
   inactivity: 300  # 5 minutes
   remember_me_duration: 1y 
```

**Notes**
* for backguard compability `session` and `sessions` sections must coexists
* requested domains must match a domain listed in `domains` (this may be being redundant with ACL)
* order doesn't matter 

let me know if i missing something, and which alternative should implement.

TY!

nightah commented:
I'm thinking sort of a hybrid of the two:

```yaml
session:
  domain: example.com #backward compatibility
  domains:
    -   cookie_domain: internal.example.com
        domains:
            - auth.internal.example.com
            - internal.example.com
        expiration: 1800
        inactivity: 300  # 5 minutes
        remember_me_duration: 3y 
    -   cookie_domain: example.com
        domains:
            - auth.example.com
            - secure.example.com
            - public.example.com
        expiration: 1800
        inactivity: 300  # 5 minutes
        remember_me_duration: 3y 
    -   cookie_domain: example2.com
        domains:
            - example2.com
            - *.example2.com
        expiration: 1800
        inactivity: 300  # 5 minutes
        remember_me_duration: 3y 
  # default values
  same_site: true
  expiration: 3600  # 1 hour
  inactivity: 300  # 5 minutes
  remember_me_duration: 1y
```

That way also for the validator we can ensure that no domain exists under separate `cookie_domains` (we can throw an error), we also can support wildcards in the domain matcher. While this would mean the first match is the one that counts (like ACLs) we can ensure that the first match is also technically the only match in our validation.

@james-d-elliott also keen to hear your thoughts and whether you agree with this approach.

mind-ar commented:
Hi, i'll close this PR so we can re-implement this feature as discussed.
